File: app/src/main/java/com/hamburgcodingschool/weather/WeatherPresenter.java
package com.hamburgcodingschool.weather;

import android.support.annotation.Nullable;

import com.hamburgcodingschool.weather.model.City;

/**
 * @author Teresa Holfeld, teresa.holfeld@gmail.com
 */
public class WeatherPresenter implements WeatherRepository.WeatherRecipient {

    private static final String ERROR_TEXT = "Sorry, something went wrong.";

    private final WeatherView weatherView;
    private final WeatherRepository weatherRepository;


    public WeatherPresenter(WeatherView weatherView, WeatherRepository weatherRepository) {
        this.weatherView = weatherView;
        this.weatherRepository = weatherRepository;
    }

    public void getWeather(String city) {
        weatherRepository.getWeather(city);
    }

    @Override
    public void receiveWeather(@Nullable City city) {
        if (city == null) {
            weatherView.setWeather(ERROR_TEXT);
        } else {
            String weather = city.getWeather().get(0).getDescription();
            weatherView.setWeather(weather);
        }
    }
}

File: app/src/main/java/com/hamburgcodingschool/weather/WeatherPresenter.kt
package com.hamburgcodingschool.weather

import com.hamburgcodingschool.weather.model.City

/**
 * @author Teresa Holfeld, teresa.holfeld@gmail.com
 */
class WeatherPresenter(private val weatherView: WeatherView, private val weatherRepository: WeatherRepository) : WeatherRepository.WeatherCallback {

    fun getWeather(city: String) {
        weatherRepository.getWeather(city)
    }

    override fun receiveWeather(city: City?) {
        if (city == null) {
            weatherView.setWeather(ERROR_TEXT)
        } else {
            val weather = city.weather[0].description
            weatherView.setWeather(weather)
        }
    }

    companion object {

        internal val ERROR_TEXT = "Sorry, something went wrong."
    }
}
